pkg/allocator/allocator.go

```diff
@@ -0,0 +1,788 @@
+func (a *Allocator) Get(ctx context.Context, key AllocatorKey) (idpool.ID, error) {
+	if id := a.mainCache.get(key.GetKey()); id != idpool.NoID {
+		return id, nil
+	}
+
+	return a.GetNoCache(ctx, key)
+}
+
+// GetNoCache returns the ID which is allocated to a key in the kvstore,
+// bypassing the local copy of allocated keys.
+func (a *Allocator) GetNoCache(ctx context.Context, key AllocatorKey) (idpool.ID, error) {
+	return a.backend.Get(ctx, key)
+}
+
+// GetByID returns the key associated with an ID. Returns nil if no key is
+// associated with the ID.
+func (a *Allocator) GetByID(id idpool.ID) (AllocatorKey, error) {
+	if key := a.mainCache.getByID(id); key != nil {
+		return key, nil
+	}
+
+	return a.backend.GetByID(id)
+}
+
+// Release releases the use of an ID associated with the provided key. After
+// the last user has released the ID, the key is removed in the KVstore and
+// the returned lastUse value is true.
+func (a *Allocator) Release(ctx context.Context, key AllocatorKey) (lastUse bool, err error) {
+	log.WithField(fieldKey, key).Info("Releasing key")
+
+	select {
+	case <-a.initialListDone:
+	case <-ctx.Done():
+		return false, fmt.Errorf("release was cancelled while waiting for initial key list to be received: %s", ctx.Err())
+	}
+
+	k := key.GetKey()
+
+	a.slaveKeysMutex.Lock()
+	defer a.slaveKeysMutex.Unlock()
+
+	// release the key locally, if it was the last use, remove the node
+	// specific value key to remove the global reference mark
+	lastUse, err = a.localKeys.release(k)
+	if err != nil {
+		return lastUse, err
+	}
+	if lastUse {
+		a.backend.Release(ctx, key)
+	}
+
+	return lastUse, err
+}
+
+// RunGC scans the kvstore for unused master keys and removes them
+func (a *Allocator) RunGC(staleKeysPrevRound map[string]uint64) (map[string]uint64, error) {
+	return a.backend.RunGC(staleKeysPrevRound)
+}
+
+func (a *Allocator) DeleteAllKeys() {
+	a.backend.DeleteAllKeys()
+}
+
+// syncLocalKeys checks the kvstore and verifies that a master key exists for
+// all locally used allocations. This will restore master keys if deleted for
+// some reason.
+func (a *Allocator) syncLocalKeys(ctx context.Context) error {
+	// Create a local copy of all local allocations to not require to hold
+	// any locks while performing kvstore operations. Local use can
+	// disappear while we perform the sync but that is fine as worst case,
+	// a master key is created for a slave key that no longer exists. The
+	// garbage collector will remove it again.
+	ids := a.localKeys.getVerifiedIDs()
+
+	for id, value := range ids {
+		if err := a.backend.UpdateKey(ctx, id, value, false); err != nil {
+			return fmt.Errorf("unable to sync key \"%s\"->\"%s\": %s", id, value, err)
+		}
+	}
+
+	return nil
+}
+
+func (a *Allocator) startLocalKeySync() {
+	go func(a *Allocator) {
+		for {
+			if err := a.syncLocalKeys(context.TODO()); err != nil {
+				log.WithError(err).Warning("Unable to run local key sync routine")
+			}
+
+			select {
```

Comment:
hmm, yeah. I was plumbing in the context and then I hit this function and it was difficult so I stopped. This was mostly because I wanted to be able to cancel the sync operation (assuming it's supported all the way down the stack, but that is a lie I think).
I just realised though, that I can do this:
```
ctx, cancel := context.WithCancel(context.TODO())
...
if err := a.syncLocalKeys(ctx); err != nil {
...
			case <-a.stopGC:
                                 cancel()
				log.Debug("Stopped master key sync routine")
				return
```

This way, stopGC is wired through to the lower levels. An alternative would be to defer the cancel. What do you think? too obtuse?

Reply:
What are we gaining here with the `cancel` call? Since AFAIU we don't have anything (resources/state) associated with this context, cancelling it is basically a no-op at this point, right?

Reply:
ok. I undid this completely.